Backport: win: add ENABLE_VIRTUAL_TERMINAL_INPUT raw tty mode

KristofferC opened the pull request:
Backports https://github.com/libuv/libuv/pull/4688, which allows bracket paste to work on Windows, tested on https://github.com/JuliaLang/julia/pull/59825.

vtjnash commented on src/win/pipe.c:
```diff
@@ -1405,7 +1414,16 @@ static int uv_pipe_write_impl(uv_loop_t* loop,
+    if (!req->event_handle) {
+      uv_fatal_error(GetLastError(), "CreateEvent");
+    }
+    req->u.io.overlapped.hEvent = (HANDLE) ((uintptr_t) req->event_handle | 1);
```
At various points below, we attempt to close this fake handle (which will fail). They should be closing the real handle (req->event_handle) instead.

vtjnash replied:
(In case you're wondering what the `| 1` means, it disables notifying of the associated IOCP when this write finishes. Since we don't have an IOCP, that'll result in a small performance boost. Ref https://www.microsoftpressstore.com/articles/article.aspx?p=2224047&seqNum=5)

martinholters replied:
How about adding that as a comment?

staticfloat replied:
The `| 1`-style idiom is pretty common in Win32 APIs, and this is used in multiple (6) places throughout the codebase.  Usually I'm in favor of documenting strange things like this, but I don't particularly want to go paste the same explanation in multiple places across a couple files.  I suppose the "proper" way to do this would be to create a function called `handle_to_hEvent()` or something, and document that.

tkelman commented:
How much of this is relevant to the upstream code base?

staticfloat commented:
Supposedly all of it I think. I'll open a PR once this is merged here.

On Thu, Jun 1, 2017, 00:18 Tony Kelman <notifications@github.com> wrote:

> How much of this is relevant to the upstream code base?
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/JuliaLang/libuv/pull/44#issuecomment-305409946>, or mute
> the thread
> <https://github.com/notifications/unsubscribe-auth/AAH_aG59y-nuVTpsKOBa757dAe2Kl0Obks5r_mW6gaJpZM4NrVrO>
> .
>
-- 

-E

ararslan commented:
Bump!

tkelman commented:
> I'll open a PR once this is merged here.

Do we have a standalone way of testing this when submitting upstream?

staticfloat commented:
No, unfortunately this doesn't manifest on their test suite, even if you redirect it, which I tried.  @vtjnash do you have any idea how we could trigger this within a test in their test suite?  I think that would be a valuable addition for them.

tkelman commented:
@vtjnash ping, we should test and upstream this

tkelman commented:
@vtjnash please don't ignore this indefinitely

vtjnash commented:
Libuv upstreaming is on my low priority list. There's a dozen or so other libuv issues in that list though, so it'll need to wait its turn.

tkelman commented:
A suggestion for how to test this is what was requested here.

tkelman commented:
@vtjnash again we need a test case for this to get upstreamed

addaleax committed:
win: add ENABLE_VIRTUAL_TERMINAL_INPUT raw tty mode (#4688)

Windows provides the `ENABLE_VIRTUAL_TERMINAL_INPUT` flag for TTY input
streams as a companion flag to `ENABLE_VIRTUAL_TERMINAL_PROCESSING`,
which libuv is already setting for TTY output streams.

Setting this flag lets the terminal emulator perform some of the
processing that libuv already currently does for input events,
but most notably enables receiving control sequences that are
otherwise entirely unavailable, e.g. for bracketed paste
(which the Node.js readline implementation added basic support for
in https://github.com/nodejs/node/commit/87af913b66eab78088acfd).

libuv currently already provides translations for key events to
control sequences, i.e. what this mode is intended to provide,
but libuv does not and cannot translate all such events.
Since the control sequences differ from the ones that Windows
has chosen to standardize on, and applications may not be expecting
this change, this is opt-in for now (but ideally will be the default
behavior starting in libuv v2.x, should that ever happen).

Another downside of this change is that not all shells reset
this mode when an application exits. For example, when running a
Node.js program with this flag enabled inside of PowerShell in
Windows terminal, if the application exits while in raw TTY input mode,
neither the shell nor the terminal emulator reset this flag, rendering
the input stream unusable.

While there's general awareness of the problem that console state is
global state rather than per-process (same as on UNIX platforms),
it seems that applications like PowerShell aren't expecting to need to
unset this flag on the input stream, only its output counterpart
(e.g. https://github.com/PowerShell/PowerShell/blob/4e7942135f998ab40fd3ae298b020e161a76d4ef/src/Microsoft.PowerShell.ConsoleHost/host/msh/ConsoleHost.cs#L1156).

Hence, `uv_tty_reset_mode()` is extended to reset the terminal
to its original state if the new mode is being used.

Refs: https://github.com/nodejs/node/commit/87af913b66eab78088acfd7f3b57d35e22c5f1ba
Refs: https://github.com/microsoft/terminal/issues/4954
(cherry picked from commit 843b64faf5fbbd19c04475bafa38c0c91514efb8)

KristofferC commented:
I don't know what the state of CI is on this repo but looking at other PRs it seems to at least not be green...
